direct/direct.go

```diff
@@ -107,65 +106,38 @@ func NewClient(keytype string, mnemonics string, relayUrl string, session string
 			return nil, errors.Wrap(err, "could not update twin relay information")
 		}
 	}
+	conn := NewConnection(identity, relayDomain, session, twin.ID)
 
-	token, err := NewJWT(identity, id, session, 60) // use 1 min token ttl
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to build authentication token")
-	}
-	// wss://relay.dev.grid.tf/?<JWT>
-	relayUrl = fmt.Sprintf("%s?%s", relayUrl, token)
+	reader, writer := conn.Start(ctx)
 	source := types.Address{
 		Twin:       id,
 		Connection: &session,
 	}
 
-	con, resp, err := websocket.DefaultDialer.Dial(relayUrl, nil)
-	if err != nil {
-		var body []byte
-		var status string
-		if resp != nil {
-			status = resp.Status
-			body, _ = io.ReadAll(resp.Body)
-		}
-		return nil, errors.Wrapf(err, "failed to connect (%s): %s", status, string(body))
-	}
-
-	if resp.StatusCode != http.StatusSwitchingProtocols {
-		return nil, fmt.Errorf("invalid response %s", resp.Status)
-	}
-
 	cl := &directClient{
 		source:    &source,
 		signer:    identity,
-		con:       con,
 		responses: make(map[string]chan *types.Envelope),
 		twinDB:    twinDB,
 		privKey:   privKey,
+		reader:    reader,
+		writer:    writer,
 	}
-
 	go cl.process()
+
 	return cl, nil
 }
-
 func (d *directClient) process() {
-	defer d.con.Close()
+	// defer d.connection.con.Close()
 	// todo: set error on connection here
 	for {
-		typ, msg, err := d.con.ReadMessage()
-		if err != nil {
-			log.Error().Err(err).Msg("websocket error connection closed")
-			return
-		}
-		if typ != websocket.BinaryMessage {
-			continue
-		}
+		incoming := <-d.reader
```

Comment:
this can be rewritten as 
```go 
for incoming := range d.reader {
```

> Note: this loop will also auto exit when the d.reader is closed when the inner connection is closing